homeassistant/components/fan/dyson.py

```diff
@@ -10,62 +10,153 @@
-        dyson_entity = DysonPureCoolLinkDevice(hass, device)
-        hass.data[DYSON_FAN_DEVICES].append(dyson_entity)
+    has_purecool_devices = False
+    device_names = (device.name for device in hass.data[DYSON_FAN_DEVICES])
+    for device in hass.data[DYSON_DEVICES]:
+        if device.name not in device_names:
+            if isinstance(device, DysonPureCool):
+                has_purecool_devices = True
+                dyson_entity = DysonPureCoolDevice(device)
+                hass.data[DYSON_FAN_DEVICES].append(dyson_entity)
+            elif isinstance(device, DysonPureCoolLink):
+                dyson_entity = DysonPureCoolLinkDevice(hass, device)
+                hass.data[DYSON_FAN_DEVICES].append(dyson_entity)
 
     add_entities(hass.data[DYSON_FAN_DEVICES])
 
     def service_handle(service):
         """Handle the Dyson services."""
-        entity_id = service.data.get(CONF_ENTITY_ID)
-        night_mode = service.data.get(CONF_NIGHT_MODE)
-        fan_device = next([fan for fan in hass.data[DYSON_FAN_DEVICES] if
-                           fan.entity_id == entity_id].__iter__(), None)
+        entity_id = service.data.get(ATTR_ENTITY_ID)
+        night_mode = service.data.get(ATTR_NIGHT_MODE)
+        fan_device = next((fan for fan in hass.data[DYSON_FAN_DEVICES] if
+                           fan.entity_id == entity_id), None)
         if fan_device is None:
             _LOGGER.warning("Unable to find Dyson fan device %s",
                             str(entity_id))
             return
 
         if service.service == SERVICE_SET_NIGHT_MODE:
-            fan_device.night_mode(night_mode)
+            fan_device.set_night_mode(night_mode)
+
+        if has_purecool_devices:
+            auto_mode = service.data.get(ATTR_AUTO_MODE)
+            angle_low = service.data.get(ATTR_ANGLE_LOW)
+            angle_high = service.data.get(ATTR_ANGLE_HIGH)
+            flow_direction_front = service.data.get(ATTR_FLOW_DIRECTION_FRONT)
+            timer = service.data.get(ATTR_TIMER)
+            dyson_speed = service.data.get(ATTR_DYSON_SPEED)
+
+            if service.service == SERVICE_SET_AUTO_MODE:
+                fan_device.set_auto_mode(auto_mode)
+
+            if service.service == SERVICE_SET_ANGLE:
+                fan_device.set_angle(angle_low, angle_high)
+
+            if service.service == SERVICE_SET_FLOW_DIRECTION_FRONT:
+                fan_device.set_flow_direction_front(flow_direction_front)
+
+            if service.service == SERVICE_SET_TIMER:
+                fan_device.set_timer(timer)
+
+            if service.service == SERVICE_SET_DYSON_SPEED:
+                fan_device.set_dyson_speed(dyson_speed)
 
     # Register dyson service(s)
     hass.services.register(
         DOMAIN, SERVICE_SET_NIGHT_MODE, service_handle,
         schema=DYSON_SET_NIGHT_MODE_SCHEMA)
 
+    if has_purecool_devices:
+        hass.services.register(
+            DOMAIN, SERVICE_SET_AUTO_MODE, service_handle,
```

Comment:
All of the dyson custom services should be registered under the dyson component "domain".

We're also embedding the platform modules under the integrating component (dyson). That's why there are merge conflicts in this PR.

Reply:
Hmm did I do something wrong? I moved them to the main component, but I still see the merge conflict message here.

Reply:
You need to rebase the branch on latest dev too.

Reply:
Makes sense. Let's see if I can do it without messing up the commit history this time.

Reply:
I guess I can't... I'll open a new pull request :).